File: app/src/main/java/com/example/testtanaw/CameraActivity.java
package com.example.testtanaw;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.view.WindowMetrics;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.AspectRatio;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.camera.core.resolutionselector.ResolutionSelector;
import androidx.camera.core.resolutionselector.AspectRatioStrategy;
import androidx.camera.core.Preview;
import com.example.testtanaw.models.Constants;
import com.example.testtanaw.models.UserSdgPhoto;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import androidx.camera.lifecycle.ProcessCameraProvider;
import com.google.common.util.concurrent.ListenableFuture;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.time.LocalDateTime;
import java.util.UUID;

public class CameraActivity extends AppCompatActivity {

    private static final String TAG = "CameraActivity";

    private FirebaseAuth mAuth;
    private FirebaseFirestore db;
    private FirebaseStorage storage;
    private FirebaseUser authUser;


    private ImageCapture imageCapture;
    private ProcessCameraProvider cameraProvider;
    private Camera camera;
    private CameraSelector cameraSelector;
    private int lensFacing = CameraSelector.LENS_FACING_BACK;
    private byte[] curImageData;


    private int sdgNumber;
    private String photoChallenge;
    private Double latitude;
    private Double longitude;

    private String path;
    private UserSdgPhoto newSdgPhoto;
    private byte[] imageData;

    private ImageButton fabCamera;
    private ImageButton fabFlipCamera;
    private EditText caption;
    private ImageButton upload;
    private ImageButton retake;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_camera);


        // Initialize Firebase instances
        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();
        storage = FirebaseStorage.getInstance(Constants.BUCKET);

        authUser = mAuth.getCurrentUser();


        if (authUser != null) {
            // get intent
            sdgNumber = getIntent().getIntExtra("SDGNUMBER", -1);
            photoChallenge = getIntent().getStringExtra("PHOTOCHALLENGE");
            latitude = getIntent().getDoubleExtra("LAT", 0.0);
            longitude = getIntent().getDoubleExtra("LONG", 0.0);

            // set views
            fabCamera = findViewById(R.id.fab_camera);
            fabFlipCamera = findViewById(R.id.fab_flip_camera);
            caption = findViewById(R.id.caption);
            upload = findViewById(R.id.upload);
            retake = findViewById(R.id.retake);

            setInitialState();

            fabCamera.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    takePhoto();
                }
            });

            upload.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    uploadPhoto();
                }
            });

            retake.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    retakePhoto();
                }
            });

            fabFlipCamera.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    flipCamera();
                }
            });

            // check cam permission
            if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                requestCameraPermission();
                startCamera();
            } else {
                startCamera();
            }
        } else {
            Toast.makeText(this, "User not logged in", Toast.LENGTH_SHORT).show();
        }
    }


    private void setInitialState() {
        fabCamera.setVisibility(View.VISIBLE);
        fabFlipCamera.setVisibility(View.VISIBLE);
        caption.setVisibility(View.GONE);
        upload.setVisibility(View.GONE);
        retake.setVisibility(View.GONE);
    }

    private void bindCameraUserCases() {
        DisplayMetrics displayMetrics = new DisplayMetrics();
        int width;
        int height;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            WindowMetrics windowMetrics = getWindowManager().getCurrentWindowMetrics();

            // Get the bounds of the display
            Rect bounds = windowMetrics.getBounds();
            width = bounds.width();
            height = bounds.height();

            Log.d("tag", "Width: " + width + ", Height: " + height);
        } else {
            // For older APIs, use the deprecated method
                    getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
            width = displayMetrics.widthPixels;
            height = displayMetrics.heightPixels;

            Log.d("tag", "Width: " + width + ", Height: " + height);
        }

        int screenAspectRatio = aspectRatio(width, height);

        // Determine the rotation
        int rotation = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) ?
                getWindowManager().getDefaultDisplay().getRotation() :
                getWindowManager().getDefaultDisplay().getRotation();

        ResolutionSelector resolutionSelector = new ResolutionSelector.Builder()
                .setAspectRatioStrategy(new AspectRatioStrategy(screenAspectRatio, AspectRatioStrategy.FALLBACK_RULE_AUTO))
                .build();

        Preview preview = new Preview.Builder()
                .setResolutionSelector(resolutionSelector)
                .setTargetRotation(rotation)
                .build();
        preview.setSurfaceProvider(((PreviewView) findViewById(R.id.cameraView)).getSurfaceProvider());

        imageCapture = new ImageCapture.Builder()
                .setCaptureMode(ImageCapture.CAPTURE_MODE_MAXIMIZE_QUALITY)
                .setResolutionSelector(resolutionSelector)
                .setTargetRotation(rotation)
                .build();

        cameraSelector = new CameraSelector.Builder()
                .requireLensFacing(lensFacing)
                .build();

        try {
            cameraProvider.unbindAll();
            camera = cameraProvider.bindToLifecycle(this, cameraSelector, preview, imageCapture);
        } catch (Exception e) {
            Log.d("tag", "Error bindcamera: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private void flipCamera() {
        Toast.makeText(this, "Camera Flipped!", Toast.LENGTH_SHORT).show();
        lensFacing = (lensFacing == CameraSelector.LENS_FACING_FRONT) ? CameraSelector.LENS_FACING_BACK : CameraSelector.LENS_FACING_FRONT;
        bindCameraUserCases();
    }

    private void takePhoto() {
        Toast.makeText(this, "Photo Captured!", Toast.LENGTH_SHORT).show();
        fabCamera.setVisibility(View.GONE);
        fabFlipCamera.setVisibility(View.GONE);
        caption.setVisibility(View.VISIBLE);
        upload.setVisibility(View.VISIBLE);
        retake.setVisibility(View.VISIBLE);

        // Implement photo capture logic here
        File imageFile = null;
        try {
            imageFile = File.createTempFile("temp", ".jpg");
        } catch (IOException e) {
            Log.d(TAG, "ERROR TAKE PHOTO FILE TEMP: " + e.getMessage());
            throw new RuntimeException(e);
        }

        ImageCapture.OutputFileOptions outputOption = new ImageCapture.OutputFileOptions.Builder(imageFile).build();
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

        imageCapture.takePicture(
                outputOption,
                ContextCompat.getMainExecutor(this),
                new ImageCapture.OnImageSavedCallback() {

                    @Override
                    public void onImageSaved(@NonNull ImageCapture.OutputFileResults outputFileResults) {
                        try {
                            Uri savedUri = outputFileResults.getSavedUri();
                            if (savedUri != null) {
                                File file = new File(savedUri.getPath());
                                try (InputStream inputStream = new FileInputStream(file)) {
                                    byte[] buffer = new byte[1024];
                                    int bytesRead;
                                    while ((bytesRead = inputStream.read(buffer)) != -1) {
                                        byteArrayOutputStream.write(buffer, 0, bytesRead);
                                    }
                                }

                                imageData = byteArrayOutputStream.toByteArray();

                                // Convert the byte array to Bitmap
                                Bitmap bitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.length);

                                // Rotate if needed (assuming you determine the required angle)
                                int angle = 90;
                                bitmap = rotateImage(bitmap, angle);

                                // Crop the Bitmap to a 1x1 (square) aspect ratio
                                bitmap = cropToSquare(bitmap);

                                // Get the ImageView from the layout
                                ImageView capturedImageView = findViewById(R.id.imgView);

                                // Set the Bitmap to the ImageView to display the captured image
                                capturedImageView.setImageBitmap(bitmap);
                            }

                        } catch (Exception e) {
                            Log.d(TAG, "ERROR TAKE PHOTO IMAGE SAVE: " + e.getMessage());
                            throw new RuntimeException(e);
                        }
                    }

                    @Override
                    public void onError(@NonNull ImageCaptureException exception) {

                    }
                });
    }

    private Bitmap rotateImage(Bitmap source, int angle) {
        Matrix matrix = new Matrix();
        matrix.postRotate(angle);
        return Bitmap.createBitmap(source, 0, 0, source.getWidth(), source.getHeight(), matrix, true);
    }

    private Bitmap cropToSquare(Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();

        int newDimension = Math.min(width, height);  // Take the smallest side for square

        // Calculate the center of the bitmap
        int x = (width - newDimension) / 2;
        int y = (height - newDimension) / 2;

        // Crop the image to a square
        Bitmap croppedBitmap = Bitmap.createBitmap(bitmap, x, y, newDimension, newDimension);

        return croppedBitmap;
    }

    private void uploadUsersSdgPhoto(UserSdgPhoto userSdgPhoto, String path) {
        if (authUser == null) {
            return;
        }

        StorageReference storageRef = storage.getReference();
        StorageReference userSdgPhotoRef = storageRef.child(path);
        Toast.makeText(CameraActivity.this, path, Toast.LENGTH_SHORT).show();

        assert imageData != null;
        UploadTask uploadTask = userSdgPhotoRef.putBytes(imageData);
        uploadTask.addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception exception) {
                Toast.makeText(CameraActivity.this, exception.getMessage(), Toast.LENGTH_SHORT).show();
                Log.e(TAG, "User sdg image upload failed.", exception);
            }
        }).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
            @Override
            public void onSuccess(UploadTask.TaskSnapshot taskSnapshot) {
                Toast.makeText(CameraActivity.this, "User Sdg image uploaded.", Toast.LENGTH_SHORT).show();
                Log.d(TAG, "User Sdg uploaded.");
            }
        });
    }

    private void saveUsersSdgPhotoDataToFirestore(UserSdgPhoto userSdgPhoto, String uuid) {
        // Save data to Firestore under the "users_sdg_photos" collection
        db.collection(Constants.DB_USERS_SDG_PHOTOS)
                .document(userSdgPhoto.getUserSdgId())
                .set(userSdgPhoto)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Toast.makeText(CameraActivity.this, "User Sdg Photo data saved.", Toast.LENGTH_SHORT).show();

//                         redirect to MainActivity
                        Intent intent = new Intent(CameraActivity.this, MainActivity.class);
                        startActivity(intent);
                    } else {
                        Toast.makeText(CameraActivity.this, task.getException().getMessage(),
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void uploadPhoto() {
        String uuid = UUID.randomUUID().toString();
        LocalDateTime now = LocalDateTime.now();

        String captionText = caption.getText().toString();

        if (authUser != null) {
            path = UserSdgPhoto.getUserSdgPhotoPath(String.valueOf(sdgNumber), uuid);
            newSdgPhoto = new UserSdgPhoto(uuid, authUser.getUid(), String.valueOf(sdgNumber), captionText, now, path, photoChallenge, latitude, longitude);
        } else {
            Toast.makeText(this, "No user is currently logged in.", Toast.LENGTH_SHORT).show();
        }


        if (!captionText.isEmpty()) {
            // Your upload logic here
            Toast.makeText(this, "Uploading photo with caption: " + captionText, Toast.LENGTH_SHORT).show();
            uploadUsersSdgPhoto(newSdgPhoto, path);
            saveUsersSdgPhotoDataToFirestore(newSdgPhoto, uuid);
        } else {
            Toast.makeText(this, "Add caption", Toast.LENGTH_SHORT).show();
        }
    }

    private void retakePhoto() {
        Toast.makeText(this, "Retaking Photo!", Toast.LENGTH_SHORT).show();

        // clear img
        ImageView capturedImageView = findViewById(R.id.imgView);
        capturedImageView.setImageBitmap(null);

        setInitialState();
    }

    // Define the permission request launcher
    private final ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), new ActivityResultCallback<Boolean>() {
                @Override
                public void onActivityResult(Boolean isGranted) {
                    if (isGranted) {
                        Log.d(TAG, "Camera opened");
                        startCamera();
                        // Do capture
                    } else {
                        Log.d(TAG, "Permission denied");
                    }
                }
            });

    // Request the camera permission
    private void requestCameraPermission() {
        if (ActivityCompat.shouldShowRequestPermissionRationale(this, android.Manifest.permission.CAMERA)) {
            // Explain to the user why we need this permission, then launch the permission request
            requestPermissionLauncher.launch(android.Manifest.permission.CAMERA);
        } else {
            // Just launch the permission request
            requestPermissionLauncher.launch(android.Manifest.permission.CAMERA);
        }
    }

    private void startCamera() {
        // Get the ProcessCameraProvider instance
        ListenableFuture<ProcessCameraProvider> listendableFuture = ProcessCameraProvider.getInstance((this));

        listendableFuture.addListener(new Runnable() {
            @Override
            public void run() {
                try {
                    // Get the cameraProvider instance
                    cameraProvider = ProcessCameraProvider.getInstance(getApplicationContext()).get();

                    // Bind the camera use cases
                    bindCameraUserCases();
                } catch (Exception e) {
                    // Log the error if the camera setup fails
                    Log.e(TAG, "Camera Error: " + e.getMessage());
                }
            }
        }, ContextCompat.getMainExecutor(this));
    }

    private int aspectRatio(int width, int height) {
        double previewRatio = Math.max(height, width) / (double) Math.min(height, width);
        if (Math.abs(previewRatio - (4.0 / 3.0)) <= Math.abs(previewRatio - (16.0 / 9.0))) {
            return AspectRatio.RATIO_4_3;
        } else {
            return AspectRatio.RATIO_16_9;
        }
    }


}

File: app/src/main/java/com/example/testtanaw/CameraActivity.kt
package com.example.testtanaw
import android.Manifest
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.util.DisplayMetrics
import android.util.Log
import android.view.Surface
import android.view.View
import android.view.WindowMetrics
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.AspectRatio
import androidx.camera.core.Camera
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCapture.OutputFileOptions
import androidx.camera.core.ImageCapture.OutputFileResults
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.Preview
import androidx.camera.core.resolutionselector.AspectRatioStrategy
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.testtanaw.models.UserParcelable
import com.example.testtanaw.util.CRUD
import com.example.testtanaw.util.DB
import io.github.jan.supabase.auth.Auth
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.realtime.Realtime
import io.github.jan.supabase.storage.Storage
import io.github.jan.supabase.storage.storage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.Date
import java.util.UUID
import kotlin.math.abs

private val supabase = createSupabaseClient(
    supabaseUrl = BuildConfig.SUPABASE_URL,
    supabaseKey = BuildConfig.SUPABASE_ANON_KEY
) {
    install(Auth)
    install(Postgrest)
    install(Realtime)
    install(Storage)
    //install other modules
}

class CameraActivity : AppCompatActivity() {

    private lateinit var imageCapture: ImageCapture
    private lateinit var cameraProvider: ProcessCameraProvider
    private lateinit var camera: Camera
    private lateinit var cameraSelector: CameraSelector
    private var lensFacing = CameraSelector.LENS_FACING_BACK
    private lateinit var curImageData: ByteArray
    val db = DB()
//    val crud = CRUD()
    var userId: String? = null
    var sdgNumber: String? = null
    var photoChallenge: String? = null
    var lat: Double? = 0.0
    var long: Double? = 0.0



    private lateinit var fabCamera: ImageButton
    private lateinit var fabFlipCamera: ImageButton
    private lateinit var caption: EditText
    private lateinit var upload: ImageButton
    private lateinit var retake: ImageButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_camera)

        Log.d("xxxxxx", "HEREEEEEE")
        CoroutineScope(Dispatchers.Main).launch {
            userId = db.getUserId()
        }

        try {
            sdgNumber = intent.getStringExtra("sdgNumber")
            photoChallenge = intent.getStringExtra("photoChallenge")
            lat = intent.getStringExtra("lat")?.toDouble()
            long = intent.getStringExtra("long")?.toDouble()


            fabCamera = findViewById(R.id.fab_camera)
            fabFlipCamera = findViewById(R.id.fab_flip_camera)
            caption = findViewById(R.id.caption)
            upload = findViewById(R.id.upload)
            retake = findViewById(R.id.retake)

            // Initial state
            setInitialState()

            // Camera FAB click listener
            fabCamera.setOnClickListener {
                takePhoto()
            }

            // Upload FAB click listener
            upload.setOnClickListener {
                uploadPhoto()
            }

            // Retake FAB click listener
            retake.setOnClickListener {
                retakePhoto()
            }

            // Flip camera FAB click listener
            fabFlipCamera.setOnClickListener {
                flipCamera()
            }


        } catch (e: Exception) {
            Log.d("xxxxxx", "HEEEEE ${e.message}")
        }




        // permission
        try {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                requestCameraPermission()
                startCamera()
                // do capture
            } else {
                Log.d("xxxxxx", "cam open permission already granted")
                startCamera()
            }
        } catch (e: Exception) {
            Log.d("xxxxxx", "error ${e.message}")
        }
    }

    private fun setInitialState() {
        fabCamera.visibility = View.VISIBLE
        fabFlipCamera.visibility = View.VISIBLE
        caption.visibility = View.GONE
        upload.visibility = View.GONE
        retake.visibility = View.GONE
    }

    private fun uploadPhoto() {
        // Logic to upload photo and caption

        if (lat != 0.0 || long != 0.0) {
            val captionText = caption.text.toString()
            Toast.makeText(this, "Uploading photo with caption: $captionText", Toast.LENGTH_SHORT).show()


            CoroutineScope(Dispatchers.Main).launch {

                val uuid = UUID.randomUUID().toString();

                val currentDate = Date(System.currentTimeMillis())  // Use milliseconds since epoch
                val fileDate = (currentDate.time / 1000).toString()

                // change sdg number
                val filename = "sdg1/${uuid}-${fileDate}.jpg"
                Log.d("xxxxxx", "filename ${filename}")
                try {
                    val bucket = supabase.storage.from("user_sdgs")
                    bucket.upload(filename, curImageData)

                    val url = supabase.storage.from("user_sdgs").publicUrl(filename)
                    Log.d("xxxxxx", "URLLLL: $url")

                    // add db

                    Log.d("xxxxxx", "$sdgNumber, $photoChallenge, $userId")

                    if (sdgNumber != null && photoChallenge != null && userId != null) {
                        supabase.from("user_sdgs").insert(
                            CRUD.UploadSdgPhoto(
                                userId=userId!!,
                                sdgNumber = "sdg${sdgNumber!!}",
                                type = "photo",
                                caption = captionText,
                                url = url,
                                filename = filename,
                                photoChallenge = photoChallenge!!,
                                lat = lat!!,
                                long = long!!,
                            )
                        )
                    } else {
                        Log.d("xxxxxx", "$sdgNumber, $photoChallenge, $userId")
                    }
                } catch(e: Exception) {
                    Log.d("xxxxxx", "error upload ${e.message}")
                }
            }

            // Navigate back to the previous activity
            finish()

        }
    }

    private fun retakePhoto() {
        // Logic to reset the camera preview
        Toast.makeText(this, "Retaking Photo!", Toast.LENGTH_SHORT).show()

        // Revert to the initial state
        setInitialState()

    }

    private fun flipCamera() {
        // Logic to flip the camera
        Toast.makeText(this, "Camera Flipped!", Toast.LENGTH_SHORT).show()

        Log.d("xxxxxx", "flip")

        lensFacing = if(lensFacing == CameraSelector.LENS_FACING_FRONT) {
            CameraSelector.LENS_FACING_BACK
        } else {
            CameraSelector.LENS_FACING_FRONT
        }
        bindCameraUserCases()
    }





    private val requestPermissionLauncher = registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
        if (isGranted) {
            Log.d("xxxxxx", "cam openn")
            startCamera()
            // do capture
        } else {
            Log.d("xxxxxx", "permission denied")
        }
    }

    private fun requestCameraPermission() {
        if(ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)) {
            requestPermissionLauncher.launch(Manifest.permission.CAMERA)
        } else {
            requestPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    private fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)
        cameraProviderFuture.addListener({
            cameraProvider = cameraProviderFuture.get()
            bindCameraUserCases()
        }, ContextCompat.getMainExecutor(this))
    }

    private fun aspectRatio(width: Int, height: Int): Int {
        val previewRatio = maxOf(width, height).toDouble() / minOf(width, height)
        return if(abs(previewRatio - 4.0 / 3.0 ) <= abs(previewRatio - 16.0 / 9.0)) {
            AspectRatio.RATIO_4_3
        } else {
            AspectRatio.RATIO_16_9
        }

    }

    private fun bindCameraUserCases() {
        val displayMetrics = DisplayMetrics()
        val width: Int
        val height: Int

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val windowMetrics: WindowMetrics = windowManager.currentWindowMetrics

            // Get the bounds of the display
            val bounds = windowMetrics.bounds
            width = bounds.width()
            height = bounds.height()

            println("Width: $width, Height: $height")
        } else {
            // For older APIs, use the deprecated method
            @Suppress("DEPRECATION")
            windowManager.defaultDisplay.getMetrics(displayMetrics)
            width = displayMetrics.widthPixels
            height = displayMetrics.heightPixels

            println("Width: $width, Height: $height")
        }

        val screenAspectRatio = aspectRatio(width, height)

        // Determine the rotation
        val rotation = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            display?.rotation ?: Surface.ROTATION_0
        } else {
            @Suppress("DEPRECATION")
            windowManager.defaultDisplay.rotation
        }

        val resolutionSelector = ResolutionSelector.Builder()
            .setAspectRatioStrategy(
                AspectRatioStrategy(
                    screenAspectRatio,
                    AspectRatioStrategy.FALLBACK_RULE_AUTO
                )
            ).build()

        val preview = Preview.Builder()
            .setResolutionSelector(resolutionSelector)
            .setTargetRotation(rotation)
            .build()
            .also {
                it.setSurfaceProvider(findViewById<androidx.camera.view.PreviewView>(R.id.cameraView).surfaceProvider)
            }

        imageCapture = ImageCapture.Builder()
            .setCaptureMode(ImageCapture.CAPTURE_MODE_MAXIMIZE_QUALITY)
            .setResolutionSelector(resolutionSelector)
            .setTargetRotation(rotation)
            .build()

        cameraSelector = CameraSelector.Builder()
            .requireLensFacing(lensFacing)
            .build()

        try {
            cameraProvider.unbindAll()
            camera = cameraProvider.bindToLifecycle(
                this, cameraSelector, preview, imageCapture
            )
        } catch (e: Exception) {
            Log.d("xxxxxx", "Error bindcamera ${e.message}")
            e.printStackTrace()
        }
    }

    private fun setFlashIcon(camera: Camera) {
        if(camera.cameraInfo.hasFlashUnit()) {
            if(camera.cameraInfo.torchState.value == 0) {
                camera.cameraControl.enableTorch(true)
            } else {
                camera.cameraControl.enableTorch(false)
            }
        } else {
            Log.d("xxxxxx", "Flash not avail")
        }
    }

    private fun takePhoto() {
        Toast.makeText(this, "Photo Captured!", Toast.LENGTH_SHORT).show()

        // Toggle visibility
        fabCamera.visibility = View.GONE
        fabFlipCamera.visibility = View.GONE
        caption.visibility = View.VISIBLE
        upload.visibility = View.VISIBLE
        retake.visibility = View.VISIBLE


        val temp = "temp"
        // Create a temporary file (image not saved to disk yet)
        val imageFile = File.createTempFile(temp, ".jpg")
        val outputOption = OutputFileOptions.Builder(imageFile).build()

        val byteArrayOutputStream = ByteArrayOutputStream()

        imageCapture.takePicture(
            outputOption,
            ContextCompat.getMainExecutor(this),
            object : ImageCapture.OnImageSavedCallback {
                override fun onImageSaved(outputFileResults: OutputFileResults) {
                    try {
                        val file = outputFileResults.savedUri?.let { File(it.path) }
                        val inputStream = file?.inputStream()
                        inputStream?.let {
                            it.copyTo(byteArrayOutputStream)
                        }

                        val imageData = byteArrayOutputStream.toByteArray()

                        // Convert the byte array to Bitmap
                        val bitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.size)

                        // Get the ImageView from the layout
                        val capturedImageView: ImageView = findViewById(R.id.imgView)

                        // Set the Bitmap to the ImageView to display the captured image
                        capturedImageView.setImageBitmap(bitmap)


                        // wait for retake initialize variable muna

                        curImageData = imageData

                    } catch (e: Exception) {
                        Log.e("xxxxxx", "Error saving or displaying image: ${e.message}")
                    }
                }

                override fun onError(exception: ImageCaptureException) {
                    Log.d("xxxxxx", "Error: Image not captured")
                }
            }
        )
    }
}
